Nodejs/Product/TestAdapter/TestFrameworkFactory.cs

```diff
@@ -0,0 +1,111 @@
+{
+    public static class TestFrameworkFactory
+    {
+        public static Dictionary<string, HashSet<string>> GetTestItems(string projectRoot, Project project)
+        {
+            var configItems = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
+            var testItems = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
+
+            var projectItems = project.Items.Where(x => x.ItemType != "NodeModulesContent");
+
+            foreach (var projectItem in projectItems)
+            {
+                string testFrameworkName = null;
+                // TODO: Consider it can also be inside a <root>/.config/ folder and will configure the <root> folder.
+                if (projectItem.EvaluatedInclude.Contains("angular.json"))
+                {
+                    testFrameworkName = "angular";
+                }
+                // TODO: Add configuration files for jest, mocha , jasmine. Consider that some frameworks can be configured using package.json
+                // Also some frameworks have more than one filename to define a configuration.
+                // Tape is the only framework we support that doesn't have a configuration file. Decide what to do about that.
+                if (testFrameworkName != null)
+                {
+                    if (!configItems.ContainsKey(testFrameworkName))
+                    {
+                        configItems.Add(testFrameworkName, new HashSet<string>(StringComparer.OrdinalIgnoreCase));
+                    }
+
+                    configItems[testFrameworkName].Add(CommonUtils.GetAbsoluteFilePath(projectRoot, projectItem.EvaluatedInclude));
+                }
+                else if (!configItems.Any()) // Legacy behavior. If we have found a test configuration file is safe to ignore the rest of the files.
+                {
+                    // TODO: Deprecate. Is configured by project file or is configured per file.
+                    var testFrameworkAndFilePath = GetTestFrameworkAndFilePath(project, projectItem, projectRoot);
+                    if (testFrameworkAndFilePath.HasValue)
+                    {
+                        var (testFramework, fileAbsolutePath) = testFrameworkAndFilePath.Value;
+                        if (!testItems.ContainsKey(testFramework))
+                        {
+                            testItems.Add(testFramework, new HashSet<string>(StringComparer.OrdinalIgnoreCase));
+                        }
+
+                        testItems[testFramework].Add(fileAbsolutePath);
+                    }
+                }
+            }
+
+            return configItems.Any() ? configItems : testItems;
+        }
+
+        private static (string, string)? GetTestFrameworkAndFilePath(Project project, ProjectItem projectItem, string projectRoot)
+        {
+            var testRoot = project.GetProperty(NodeProjectProperty.TestRoot)?.EvaluatedValue;
```

Comment:
Do we really need to make these two calls for every item in a project?

Reply:
We do not. Nonetheless I feel like placing them inside the method makes sense as they are part of the `project` variable. Otherwise we'll need to extract them on the caller and send them as a parameter.

Is your concern based on performance? I don't think it should be a problem, although I admit I don't know much of `GetProperty` to have an argument.